soc: arm: efr32bg22: Add support for Gecko SPI

fkokosinski opened the pull request:
This PR adds support for the Gecko SPI and adds an SPI transaction sample in `samples/drivers/spi_efr32bg22/src/spi.c`.

lpawelcz committed:
dts/boards: efr32bg_sltb010a: add spi-through-usart nodes

This commit configures USART0 to act as SPI.

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz committed:
drivers: pinctrl: silabs: add spi handling

This commit adds pinctrl configuration for SPI on USART.

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz committed:
drivers: spi: gecko: add support for pinctrl configuration

This commit adds support for pinctrl configuration in the Gecko SPI
driver.

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz committed:
boards: arm: efr32bg: dts: add spi_enable gpio

This commit adds spi_enable to the efr32bg_sltb010a board dts.

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz committed:
drivers: spi: gecko: add support for CPOL and CPHA

This commit adds support for CPOL/CPHA configuration in the Gecko SPI
driver.

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz committed:
dts: spi: silabs: make peripheral-id property optional

This commit makes the peripheral-id property optional and removes it's
usage from the Gecko SPI driver.

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

fkokosinski committed:
soc: arm :efr32bg22: depend on DT_HAS_*_ENABLED for SPI selection

This commit modifies the defconfig for efr32bg22 SoC so that Gecko SPI
selection depends on the DT_HAS_*_ENABLED define.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

gmarull commented on samples/drivers/spi_efr32bg22/src/spi.c:
```diff
@@ -0,0 +1,135 @@
+{
+	LOG_INF("SPI test on buffers TX/RX %p / %p", (void *)buffer_tx, (void *)buffer_rx);
+
+	int ret;
+
+	LOG_INF("Configure SPI Enable GPIO as output");
+	ret = gpio_pin_configure_dt(&spi_enable, GPIO_OUTPUT);
+	if (ret != 0) {
+		LOG_ERR("Failed to configure SPI enable GPIO as output, return code: %d", ret);
+	}
+
+	LOG_INF("Enable SPI bus on board");
+	ret = gpio_pin_set_dt(&spi_enable, 1);
+	if (ret != 0) {
+		LOG_ERR("Failed to enable SPI bus, return code: %d", ret);
+	}
+
+	LOG_INF("Check if SPI is ready");
+	if (!spi_is_ready(&spi_efr32bg22)) {
+		LOG_ERR("SPI device is not ready");
+	}
+
+	LOG_INF("Read ID from MX25R SPI Flash device");
+	if (spi_read_rdid(&spi_efr32bg22)) {
+		return;
+	}
+
+	LOG_INF("Disable SPI bus on board");
+	ret = gpio_pin_set_dt(&spi_enable, 0);
+	if (ret != 0) {
+		LOG_ERR("Failed to disable SPI bus, return code: %d", ret);
+	}
+
+	LOG_INF("Done");
+}
+
+void main(void)
+{
+	test_spi();
+}
```
-1 on platform specific samples for interfaces. Also, the SPI enable bit should be part of the SPI driver and be specified in the SPI bindings.

lpawelcz replied:
Removed the sample. Please note that the SPI enable bit was not used as CS line. It was used for driving on-board switch. The switch connected SPI peripheral used in initial SPI driver tests to the common SPI bus on the platform. This should be handled by users in their own applications so I also removed this part from the PR.
FYI: I'm pasting here a figure from Thunderboard EFR32BG22 User Guide that highlights the usage of SPI enable bit (EFR32_IMU_ENABLE)
![image](https://user-images.githubusercontent.com/33930985/209128754-1320465f-877c-43ea-8fa0-0516e698161b.png)

gmarull replied:
note that on-board switch should either use the `regulator-fixed` compatible, or, if specific to a device you can also use `-gpios` prop to directly toggle the GPIO in the driver (probably combined with PM).

lpawelcz replied:
I'm not sure if I understand that correctly. Would you like me to reintroduce the SPI enable bit but using existing compatible or the `-gpios` prop? It is specific to one board and one sensor on that board. That is also why I don't think it should be handled by SPI driver. In my opinion this should be handled on application level when user decides to actually power up and use the sensor. Please let me know if I can leave the SPI enable pin removed or if I should rework it.

gmarull replied:
If that's the case, you can define a `fixed-regulator` for that GPIO. Then clients can use the `vin-supply` property to capture the regulator handle and use `regulator_enable/disable` as needed. However, usage of regulators is rarely supported in subsystems, I hope this improves in the future.

lpawelcz replied:
Added in https://github.com/zephyrproject-rtos/zephyr/pull/53255/commits/7402a6bb6f9fc6a2aa6998f73590f71a957a4c57.

gmarull commented on dts/arm/silabs/efr32bg22-pinctrl.dtsi:
```diff
@@ -10,10 +10,9 @@
-			psels = <GECKO_PSEL(UART_TX, E, 10)>,
-				<GECKO_PSEL(UART_RX, E, 11)>,
-				<GECKO_LOC(UART, 0)>;
+			psels = <GECKO_PSEL(SPI_SCK, C, 2)>,
+				<GECKO_PSEL(SPI_MOSI, C, 0)>,
+				<GECKO_PSEL(SPI_MISO, C, 1)>;
```
non-blocking: maybe a comment saying UART is operating as SPI here.

lpawelcz replied:
Added comments

gmarull commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -54,11 +54,31 @@
+
+			cs-gpios = <&gpioc 3 GPIO_ACTIVE_LOW>;
+
+			status = "disabled";
+			mx25r80: mx25r8035f@0 {
+				compatible = "jedec,spi-nor";
+				reg = <0>;
+				spi-max-frequency = <80000000>;
+				size = <0x800000>;
+				jedec-id = [c2 28 14];
+				sfdp-bfp = [
+					 e5 20 f1 ff  ff ff 7f 00  44 eb 08 6b  08 3b 04 bb
+					 ee ff ff ff  ff ff 00 ff  ff ff 00 ff  0c 20 0f 52
+					 10 d8 00 ff  23 72 f5 00  82 ed 04 b7  44 83 38 44
+					 30 b0 30 b0  f7 c4 d5 5c  00 be 29 ff  f0 d0 ff ff
+				];
```
is this flash embedded in the SoC?

lpawelcz replied:
It is a separate on-board chip, I guess I should move this to board-specific dts then

lpawelcz replied:
moved `mx25r80` to `boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts`

gmarull commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -54,11 +54,18 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			cs-gpios = <&gpioc 3 GPIO_ACTIVE_LOW>;
```
is this GPIO connected internally at SoC level?

kgugala replied:
this refers to SoC level GPIO controller

gmarull replied:
this refers to a specific GPIO pin, not only the controller. Is it hard-wired in the SoC?

lpawelcz replied:
CS is connected at the board level - moved `cs-gpios` to board-specific DTS.

gmarull reviewed (dismissed):
one comment

MaureenHelm reviewed (changes requested):
You should add spi to the board yaml and doc, but otherwise LGTM

lpawelcz committed:
boards: arm: efr32bg: list SPI in documentation

Add entry for SPI in efr32bg_sltb010a documentation and twister metadata
yaml file.

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz commented:
@MaureenHelm I've listed SPI in the docs and in the yaml file with Twister metadata

lpawelcz commented:
Hi @MaureenHelm @gmarull, sorry for pinging you again but I had to rebase this PR to solve conflicts created by merging [I2C support](https://github.com/zephyrproject-rtos/zephyr/pull/53246) PR. Please take another look at this PR.
